src/app/games/diam/diam.component.ts

```diff
@@ -0,0 +1,320 @@
+            return MGPValidation.SUCCESS;
+        } else {
+            return this.cancelMove(RulesFailure.MUST_CHOOSE_PLAYER_PIECE());
+        }
+    }
+    public async onRemainingPieceClick(piece: DiamPiece): Promise<MGPValidation> {
+        const clickValidity: MGPValidation = this.canUserPlay('#piece_' + piece.owner.value + '_' + (piece.otherPieceType ? 1 : 0));
+        if (clickValidity.isFailure()) {
+            return this.cancelMove(clickValidity.getReason());
+        }
+
+        if (piece.owner === this.rules.node.gameState.getCurrentPlayer()) {
+            this.selected = { type: 'pieceFromReserve', piece };
+            this.updateViewInfo();
+            return MGPValidation.SUCCESS;
+        } else {
+            return this.cancelMove(RulesFailure.MUST_CHOOSE_PLAYER_PIECE());
+        }
+    }
+    public updateBoard(): void {
+        const lastMove: DiamMove | null = this.rules.node.move;
+        if (lastMove != null) {
+            const previousState: DiamState = this.rules.node.mother.gameState;
+            if (lastMove.isDrop()) {
+                this.lastMoved = this.getLastMovedFromDrop(lastMove, previousState);
+            } else {
+                this.lastMoved = this.getLastMovedFromShift(lastMove, previousState);
+            }
+        }
+        this.updateViewInfo();
+    }
+    private getLastMovedFromDrop(drop: DiamMoveDrop, stateBefore: DiamState): LastMoved[] {
+        const startDrawPosition: Coord =
+            this.getDrawPositionRemainingPiece(drop.piece, stateBefore.getRemainingPiecesOf(drop.piece)-1);
+        const x: number = drop.getTarget();
+        const y: number = 3 - stateBefore.getStackHeight(drop.getTarget());
+        const endDrawPosition: Coord = this.getDrawPositionOnBoard(x, y);
+        const end: Coord = new Coord(x, y);
+        return [{
+            start: MGPOptional.empty(), end,
+            startDrawPosition, endDrawPosition,
+        }];
+    }
+    private getLastMovedFromShift(shift: DiamMoveShift, stateBefore: DiamState): LastMoved[] {
+        const sourceX: number = shift.start.x;
+        const destinationX: number = shift.getTarget();
+        let destinationY: number = 3 - stateBefore.getStackHeight(destinationX);
+        const lastMoved: LastMoved[] = [];
+        for (let sourceY: number = shift.start.y;
+            sourceY >= 0 && stateBefore.getPieceAtXY(sourceX, sourceY) !== DiamPiece.EMPTY;
+            sourceY--, destinationY--) {
+            lastMoved.push({
+                start: MGPOptional.of(new Coord(sourceX, sourceY)),
+                end: new Coord(destinationX, destinationY),
+                startDrawPosition: this.getDrawPositionOnBoard(sourceX, sourceY),
+                endDrawPosition: this.getDrawPositionOnBoard(destinationX, destinationY),
+            });
+        }
+        return lastMoved;
+    }
+    private updateViewInfo(): void {
+        for (let x: number = 0; x < DiamState.WIDTH; x++) {
```

Comment:
prefer function less than 20 lines as well :p ?
Even if that's stricter than 20 branch :p

Reply:
euh oui très stricter :D Mais j'ai fortement réduit via fonctions !
Y'a une règle eslint qu'on pourrait activer, mais mettre la limite à 20 nous rajoute 830 warnings donc faudrait commencer plus haut :D

Reply:
Quand les uncovered branch seront sous les 10, on pourra commencer par attaquer les 100 les plus longues, p-e commencer par "pas plus de 40 lignes" :D (ça en fait combien ça ? )

Reply:
40+ lignes = 251 fonctions
60+ = 156
80+ = 117

Le problème c'est que là dedans, il compte les `describe(() => {})` comme des fonctions bien évidemment :D